scipy/integrate/tests/test_cubature.py

```diff
@@ -0,0 +1,719 @@
+    return (-2)**dim * 1/np.prod(alphas, axis=0).T \
+        * np.cos(2*np.pi*r.T + alphas.T @ (a+b)/2) \
+        * np.prod(np.sin(alphas.T * (a-b)/2), axis=-1)
+
+
+def genz_malik_1980_f_2(x, alphas, betas):
+    r"""
+    `f_2` from Genz and Malik 1980.
+
+    .. math:: f_2(\mathbf x) = \prod^n_{i = 1} (\alpha_i^2 + (x_i - \beta_i)^2)^{-1}
+
+    .. code-block:: mathematica
+
+        genzMalik1980f2[x_List, alphas_List, betas_List] :=
+            1/Times @@ ((alphas^2 + (x - betas)^2))
+    """
+    dim = x.shape[0]
+    num_eval_points = x.shape[-1]
+
+    # Add extra dimension to parameters for eval_points
+    alphas_reshaped = np.expand_dims(alphas, -1)
+    betas_reshaped = np.expand_dims(betas, -1)
+
+    x_reshaped = x.reshape(dim, *([1]*(len(alphas.shape) - 1)), num_eval_points)
+
+    return 1/np.prod(alphas_reshaped**2 + (x_reshaped-betas_reshaped)**2, axis=0)
+
+
+def genz_malik_1980_f_2_exact(dim, a, b, alphas, betas):
+    # Expand a and b to fit the dimension of the parameters
+    a = a.reshape(dim, *([1]*(len(alphas.shape) - 1)))
+    b = b.reshape(dim, *([1]*(len(alphas.shape) - 1)))
+
+    return (-1)**dim * 1/np.prod(alphas, axis=0) * np.prod(
+        np.arctan((a - betas)/alphas) - np.arctan((b - betas)/alphas),
+        axis=0
+    )
+
+
+def genz_malik_1980_f_3(x, alphas):
+    r"""
+    `f_3` from Genz and Malik 1980.
+
+    .. math:: f_3(\mathbf x) = \exp\left(\sum^n_{i = 1} \alpha_i x_i\right)
+
+    .. code-block:: mathematica
+
+        genzMalik1980f3[x_List, alphas_List] := Exp[Dot[x, alphas]]
+    """
+
+    dim = x.shape[0]
+    num_eval_points = x.shape[-1]
+
+    # Add extra dimension to parameters for eval_points
+    alphas_reshaped = np.expand_dims(alphas, -1)
+
+    x_reshaped = x.reshape(dim, *([1]*(len(alphas.shape) - 1)), num_eval_points)
+
+    return np.exp(np.sum(alphas_reshaped * x_reshaped, axis=0))
+
+
+def genz_malik_1980_f_3_exact(dim, a, b, alphas):
+    # Expand a and b to fit the dimension of the parameters
+    a = a.reshape(dim, *([1]*(len(alphas.shape) - 1)))
+    b = b.reshape(dim, *([1]*(len(alphas.shape) - 1)))
+
+    return (-1)**dim * 1/np.prod(alphas, axis=0) * np.prod(
+        np.exp(alphas * a) - np.exp(alphas * b),
+        axis=0
+    )
+
+
+def genz_malik_1980_f_4(x, alphas):
+    dim = x.shape[0]
+    num_eval_points = x.shape[-1]
+    alphas_reshaped = np.expand_dims(alphas, -1)
+    x_reshaped = x.reshape(dim, *([1]*(len(alphas.shape) - 1)), num_eval_points)
+
+    return ((1 + np.sum(alphas_reshaped * x_reshaped, axis=0))**(-x.shape[0]-1))
+
+
+def eval_indefinite_integral(F, a, b):
+    # TODO: find a more elegant way than transposing twice
+    points = _cartesian_product(np.array([a, b]).T).T
+    res = 0
+
+    for i, point in enumerate(points):
+        # Add up evaluations of F with correct sign at corners of hyperrectangle
+        zeroes = bin(i).count('1')
+        sign = (-1)**zeroes
+        res += sign * F(point.reshape(-1, 1))
+
+    return res
```

Comment:
This is the right way to do it, perhaps we could probably do it more elegantly. I'm posting an equivalent function I wrote that might be of use:

```

def limsum(func, a, b):
    ndim = len(a)

    lims = xp.stack([a, b], axis = 0)

    out = 0
    for ind in itertools.product(range(2), repeat = ndim):
        out += pow(-1, sum(ind) + ndim) * func(lims[ind, tuple(range(ndim))])

    return out
   ```